// @contentful/changeset-utils/dist/patch/create-patched-document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPatchedDocument = exports.CreatePatchDocumentError = void 0;
const fast_json_patch_1 = require("fast-json-patch");
const helpers_1 = require("fast-json-patch/commonjs/helpers");
class CreatePatchDocumentError extends Error {
    constructor(error, message, patch, document) {
        super(message);
        this.patch = patch;
        this.document = document;
        this.name = error.name;
        this.cause = error;
    }
}
exports.CreatePatchDocumentError = CreatePatchDocumentError;
/*
 * This function is a wrapper around fast-json-patch's applyPatch function.
 * @returns the patched document without mutating the input
 * @throws a CreatePatchDocumentError if the patch cannot be applied
 */
function createPatchedDocument(document, patch) {
    try {
        return (0, fast_json_patch_1.applyPatch)(document, patch, true, false).newDocument;
    }
    catch (error) {
        if (error instanceof helpers_1.PatchError) {
            const message = error.message.split('\n')[0];
            // Allows for grouping by a non parameterized error message
            throw new CreatePatchDocumentError(error, message, JSON.stringify(patch, null, 2), JSON.stringify(document, null, 2));
        }
        else {
            throw error;
        }
    }
}
exports.createPatchedDocument = createPatchedDocument;
