// @contentful/changeset-utils/dist/helpers/create-default-editor-interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultFieldControls = exports.createDefaultEditorInterface = void 0;
function createDefaultEditorInterface(editorInterface) {
    return {
        sys: {
            type: 'EditorInterface',
            contentType: {
                sys: {
                    id: editorInterface.sys.id,
                    linkType: 'ContentType',
                    type: 'Link',
                },
            },
            environment: editorInterface.sys.environment,
            space: editorInterface.sys.space,
            createdAt: editorInterface.sys.createdAt,
            updatedAt: editorInterface.sys.createdAt,
            id: editorInterface.sys.id,
            version: 0,
        },
        controls: editorInterface.controls?.map((control) => ({ fieldId: control.fieldId })) || [],
    };
}
exports.createDefaultEditorInterface = createDefaultEditorInterface;
function getDefaultFieldControls(fieldControls) {
    return fieldControls.map((control) => ({ fieldId: control.fieldId }));
}
exports.getDefaultFieldControls = getDefaultFieldControls;
